src/common/TrackedOp.cc

```diff
@@ -22,16 +22,69 @@ static ostream& _prefix(std::ostream* _dout)
   return *_dout << "-- op tracker -- ";
 }
 
+void OpHistoryServiceThread::BreakThread() {
+  queue_spinlock.lock();
+  _external_queue.clear();
+  _break_thread = true;
+  queue_spinlock.unlock();
+}
+
+void OpHistoryServiceThread::InsertOp(utime_t& now, TrackedOpRef op) {
+  queue_spinlock.lock();
+  _external_queue.emplace_back(now, op);
+  queue_spinlock.unlock();
+}
+
+void* OpHistoryServiceThread::entry() {
+  int sleep_time = 1000;
+  list<pair<utime_t, TrackedOpRef>> internal_queue;
+  while (1) {
+    queue_spinlock.lock();
+    if (_break_thread) {
+      queue_spinlock.unlock();
+      break;
+    }
+    internal_queue.swap(_external_queue);
+    queue_spinlock.unlock();
+    if (internal_queue.empty()) {
+      usleep(sleep_time);
+      if (sleep_time < 128000) {
+        sleep_time <<= 2;
+      }
+    } else {
+      sleep_time = 1000;
+    }
+
+    while (!internal_queue.empty()) {
+      pair<utime_t, TrackedOpRef> op = internal_queue.front();
+      _ophistory->_insert_delayed(op.first, op.second);
+      internal_queue.pop_front();
+    }
+  }
+  return nullptr;
+}
+
+
 void OpHistory::on_shutdown()
 {
+  opsvc.BreakThread();
+  opsvc.join();
   Mutex::Locker history_lock(ops_history_lock);
   arrived.clear();
   duration.clear();
   slow_op.clear();
   shutdown = true;
 }
 
-void OpHistory::insert(utime_t now, TrackedOpRef op)
+void OpHistory::insert(utime_t& now, TrackedOpRef op)
+{
+  if (shutdown)
+    return;
+
+  opsvc.InsertOp(now, op);
+}
+
+void OpHistory::_insert_delayed(utime_t& now, TrackedOpRef op)
```

Comment:
Can we come up with a better name than _insert_delayed(), given that it's now *finished* being delayed? I think I'd prefer even just _insert() if nothing better suggests itself.